mgmtfn/k8splugin/driver_test.go

```diff
@@ -0,0 +1,189 @@
+*/
+
+package k8splugin
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"runtime"
+	"testing"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+)
+
+func TestNetSetup(t *testing.T) {
+	la := netlink.NewLinkAttrs()
+	la.Name = "testlinkfoo"
+	newName := "testlinknewname"
+	address := "192.168.68.68/24"
+	defGW := "192.168.68.1"
+	staticRoute := "192.168.32.0/24"
+
+	runtime.LockOSThread()
+	locked := true
+	defer func() {
+		if locked {
+			runtime.UnlockOSThread()
+		}
+	}()
+
+	globalNS, err := netns.Get()
+	if err != nil {
+		t.Fatalf("failed to get the global network namespace: %v", err)
+	}
+	defer func() {
+		netns.Set(globalNS)
+		globalNS.Close()
+	}()
+
+	newNS, err := netns.New()
+	if err != nil {
+		t.Fatal("failed to create new network namespace")
+	}
+	defer newNS.Close()
+
+	cmd := exec.Command("sleep", "9999")
+	if err = cmd.Start(); err != nil {
+		t.Fatalf("failed to start the 'sleep 9999' process: %v", err)
+	}
+	defer cmd.Process.Kill()
+
+	pid := cmd.Process.Pid
+
+	if err = netns.Set(globalNS); err != nil {
+		t.Fatalf("failed to return to the global netns: %v", err)
+	}
+
+	// the rest of the code should run without a locked thread
+	if locked {
+		runtime.UnlockOSThread()
+		locked = false
+	}
+
+	dummy := &netlink.Dummy{LinkAttrs: la}
+	if err := netlink.LinkAdd(dummy); err != nil {
+		t.Fatalf("failed to add dummy interface: %v", err)
+	}
+
+	link, err := netlink.LinkByName(la.Name)
+	if err != nil {
+		t.Fatalf("failed to get interface by name: %v", err)
+	}
+	defer netlink.LinkDel(link)
+
+	netIf, err := net.InterfaceByName(la.Name)
+	if err != nil {
+		t.Fatalf("InterfaceByName failed: %v", err)
+	}
+
+	if netIf.Flags&net.FlagUp != 0 {
+		t.Fatalf("expected interface to be down, but it's up")
+	}
+
+	if err = setIfAttrs(pid, la.Name, address, newName); err != nil {
+		t.Fatalf("setIfAttrs failed: %v", err)
+	}
+
+	if err = setDefGw(pid, defGW, newName); err != nil {
+		t.Fatalf("setDefGw failed: %v", err)
+	}
+
+	if err = addStaticRoute(pid, staticRoute, newName); err != nil {
+		t.Fatalf("addStaticRoute failed: %v", err)
+	}
+
+	// check if the interface still has its old name & is in globalNS
+	if _, err = netlink.LinkByName(la.Name); err == nil {
+		t.Fatal("interface wasn't moved to the namespace")
+	}
+
+	// check if the interface has been renamed & is in globalNS
+	if _, err = netlink.LinkByName(newName); err == nil {
+		t.Fatal("interface wasn't moved to the namespace")
+	}
+
+	if err = netns.Set(newNS); err != nil {
+		t.Fatalf("failed to enter the new network namespace: %v", err)
+	}
+
+	// ensure the interface's name has been changed
+	newLink, err := netlink.LinkByName(newName)
+	if err != nil {
+		t.Fatalf("failed to get interface by name: %v", err)
+	}
+	defer netlink.LinkDel(newLink)
+
+	// ensure that the interface's IP address has been set properly
+	addresses, err := netlink.AddrList(newLink, netlink.FAMILY_V4)
+	ifAddr := fmt.Sprintf("%v", addresses[0].IPNet)
```

Comment:
couldn't these `%v` prints just be `.String` calls? I presume they implement `fmt.Stringer`